# ELATUS: DEMO dataset with 1k mouse brain dataset

For this particular example the mouse brain dataset consisting on 1000 cells that was analyzed in the paper is used ("Mm_brain_1k"). In this example, ELATUS uses as input the raw scRNA-seq count matrices generated by Kallisto and Cell Ranger. Next, it removes empty droplets and filters low-quality cells. After a normalization step, it clusters the cells and identifies highly-expressed lncRNAs. It retains not only highly-expressed lncRNAs robustly detected by Cell Ranger and Kallisto, but also lncRNAs that are only detected by Kallisto and whose functionality has been independently validated or that exhibit characteristics of functional lncRNAs. 

ELATUS function on this dataset could be run:
```{r}
library("ELATUS")
functional_lncRNAs <- ELATUS(kallisto_path=system.file("extdata", "kallisto_example_raw_matrix", package = "ELATUS"), kallisto_name="cells_genes_NO_multimapping", cellRanger_path=system.file("extdata", "cellRanger_example_raw_matrix", package = "ELATUS"), organism = "Mouse", lower_emptydrops = 1000, EmptyDrops_FDR_thres = 0.01, cells_mito_threshold= 15, cells_max_threshold = 30000, cells_min_genes_detected_threshold = 500, threshold_minumun_gene_counts = 250, threshold_cells_detected = 25, dimred_clustering = "PCA", k_neighbors = 5, ratio_threshold = 40, CR_threshold = 10, SI_threshold = 0.15)

head(functional_lncRNAs,5)
        candidates       ratio                 gene kallisto_total_expression cellRanger_total_expression         SI cell_type_SI crispr_intersection         category
Gm6209      Gm6209 219.7538052 ENSMUSG00000102715.2                  218.7538                     0.00000 0.23788233            7               FALSE Exclusive_lncRNA
Gm15637    Gm15637  96.4384075 ENSMUSG00000087386.2                  289.7927                     2.01532 0.20469014           15               FALSE Exclusive_lncRNA
Gm19938    Gm19938   0.9987093 ENSMUSG00000102331.2                  623.9760                   624.78372 0.13784393           11                  NA    Common_lncRNA
Snhg6        Snhg6   0.9946959 ENSMUSG00000098234.8                 1168.3728                  1174.60826 0.04570088            7                  NA    Common_lncRNA
Gm16152    Gm16152   1.4073071 ENSMUSG00000087131.8                  378.7060                   268.81030 0.14565333            3                  NA    Common_lncRNA
```


## ELATUS step-by-step pipeline
Here, we detailed the results from the different steps in the ELATUS pipeline that are run internally using the previous demo dataset.
### 1. Load input data
ELATUS first load raw count data using import_kallisto_sc and import_CellRanger_sc [functions](https://github.com/kikegoni/ELATUS/blob/main/R/import_count_matrices.R). A tutorial for generating the input data is generated is available [here](https://github.com/kikegoni/ELATUS/blob/main/demo_CellRanger_Kallisto.sh)
```{r}
kallisto <- import_kallisto_sc(kallisto_path, kallisto_name)
kallisto <- Seurat::as.SingleCellExperiment(kallisto)
kallisto_sce <- qc_metrics(kallisto, mitochondrial_ens_ids)
kallisto_sce

cellRanger <- import_CellRanger_sc(cellRanger_path)
cellRanger <- Seurat::as.SingleCellExperiment(cellRanger)
cellRanger_sce <- qc_metrics(cellRanger, mitochondrial_ens_ids)
cellRanger_sce
```

### 2. Filtering and normalization
Next, ELATUS filters empty droplets, cells with high mitochondrial content and possible multiplets, with the following recommended thresholds:
```{r}
kallisto_filt_sce_ed <- emptydrops_filt(kallisto_sce, lower = lower_emptydrops, EmptyDrops_FDR_thres = EmptyDrops_FDR_thres)
cellRanger_filt_sce_ed <- emptydrops_filt(cellRanger_sce, lower = lower_emptydrops, EmptyDrops_FDR_thres = EmptyDrops_FDR_thres)

# Remove doublets
kallisto_filt_sce_ed_nodoubs <- remove_doublets(kallisto_filt_sce_ed)
kallisto_filt_sce_ed_nodoubs <- kallisto_filt_sce_ed_nodoubs[,kallisto_filt_sce_ed_nodoubs$isDoublet == F]
cellRanger_filt_sce_ed_nodoubs <- remove_doublets(cellRanger_filt_sce_ed)
cellRanger_filt_sce_ed_nodoubs <- cellRanger_filt_sce_ed_nodoubs[,cellRanger_filt_sce_ed_nodoubs$isDoublet == F]

# Filtering (The following thresholds have been used. cells_mito_threshold=15, cells_max_threshold= 50000, cells_min_genes_detected_threshold = 500)cells_mito_threshold, cells_max_threshold, cells_min_genes_detected_threshold
kallisto_filt_sce <- Filtering(kallisto_filt_sce_ed_nodoubs,  cells_mito_threshold=cells_mito_threshold, cells_max_threshold= cells_max_threshold, cells_min_genes_detected_threshold = cells_min_genes_detected_threshold)
kallisto_filt_sce <- scuttle::logNormCounts(kallisto_filt_sce)
cellRanger_filt_sce <- Filtering(cellRanger_filt_sce_ed_nodoubs,  cells_mito_threshold=cells_mito_threshold, cells_max_threshold= cells_max_threshold, cells_min_genes_detected_threshold = cells_min_genes_detected_threshold)
cellRanger_filt_sce <- scuttle::logNormCounts(cellRanger_filt_sce)

gene_name <- gtf$gene_name[match(rownames(kallisto_filt_sce),gtf$gene_id)]
rownames(kallisto_filt_sce) <- scuttle::uniquifyFeatureNames(rownames(kallisto_filt_sce), gene_name)
gene_name <- gtf$gene_name[match(rownames(cellRanger_filt_sce),gtf$gene_id)]
rownames(cellRanger_filt_sce) <- scuttle::uniquifyFeatureNames(rownames(cellRanger_filt_sce), gene_name)
kallisto_filt_sce
cellRanger_filt_sce
```

### 3. Get highly-expressed lncRNAs
Highly-expressed lncRNAs robustly detected by Cell Ranger and Kallisto.
```{r}
candidate_lncRNAs_common <- get_candidates(kallisto_filt_sce, cellRanger_filt_sce, threshold_minumun_gene_counts = threshold_minumun_gene_counts, threshold_cells_detected = threshold_cells_detected,lncrna_names = lncrna_names,gtf=gtf, exclusive = F)
head(candidate_lncRNAs_common)
```
Highly-expressed lncRNAs exclusively detected by Kallisto
```{r}
candidate_lncRNAs_exclusive <- get_candidates(kallisto_filt_sce, cellRanger_filt_sce , threshold_minumun_gene_counts = threshold_minumun_gene_counts, threshold_cells_detected = threshold_cells_detected,lncrna_names = lncrna_names,gtf=gtf)
head(candidate_lncRNAs_exclusive)
```

### 4. Filter exclusively detected lncRNAs
Filter highly-expressed lncRNAs exclusively detected by Kallisto that are 1) independently proved as functional (hits in CRISPRi screenings) 2) exhibit characteristics of functional lncRNAs according to defined thresholds of expression and specificity.
```{r}
# clustering to calculate specificity in determined clusters
set.seed(100100100)
kallisto_filt_sce <- scater::runPCA(kallisto_filt_sce) 
g <- scran::buildSNNGraph(kallisto_filt_sce, use.dimred = dimred_clustering, k = k_neighbors ) # k is the number of nearest neighbors used to construct the graph. Choose a smaller k to more but smaller clusters as lncRNAs tend to be expressed in small subpopulations. (in this example, k=5). dimred_clustering is the dimensionality reduction (PCA here, but could be the corrected space after integrating samples)
clust <- igraph::cluster_louvain(g)$membership
print(table(clust))
kallisto_filt_sce$louvain_clusters <- factor(clust)

# Calculate the Specificity Index for each gene
SI <- SI(kallisto_filt_sce,group_by="louvain_clusters", average_by="mean")
cell_type_specific_score <- SI[["cell_type_specificity_score"]]
counts_cell_specificity_index <- SI[["counts_cell_specificity_index"]]
candidate_lncRNAs_exclusive$SI <- cell_type_specific_score[rownames(candidate_lncRNAs_exclusive)]
candidate_lncRNAs_common$SI <- cell_type_specific_score[rownames(candidate_lncRNAs_common)]
# To know in which cluster the SI is the highest 
candidate_lncRNAs_exclusive$cell_type_SI <- colnames(counts_cell_specificity_index[rownames(candidate_lncRNAs_exclusive),])[apply(counts_cell_specificity_index[rownames(candidate_lncRNAs_exclusive),],1,which.max)]
candidate_lncRNAs_exclusive <- crispr_info(crispr_data, candidate_lncRNAs_exclusive)
candidate_lncRNAs_common$cell_type_SI <- candidate_lncRNAs_common$cell_type_SI <- colnames(counts_cell_specificity_index[rownames(candidate_lncRNAs_common),])[apply(counts_cell_specificity_index[rownames(candidate_lncRNAs_common),],1,which.max)]

# Get the biologically relevant lncRNAs from these candidates (In the paper we used the following parameters: ratio_threshold = 40, CR_threshold = 10, SI_threshold = 0.15)
exclusive_lncRNAs_CRISPRi <- candidate_lncRNAs_exclusive[candidate_lncRNAs_exclusive$crispr_intersection == T,]
exclusive_biologically_relevant_lncRNAs <- biologically_relevant_lncRNAs(candidate_lncRNAs_exclusive, ratio_threshold,CR_threshold,SI_threshold)
head(exclusive_biologically_relevant_lncRNAs)
```

### 5. Genate the collection of all functional lncRNAs in scRNA-seq
This is the output of ELATUS.
```{r}
candidate_lncRNAs_common$crispr_intersection = "NA"
if (nrow(exclusive_lncRNAs_CRISPRi)>0)
{
    exclusive_lncRNAs_CRISPRi$category = "Exclusive_lncRNA_CRISPRi"
}
exclusive_biologically_relevant_lncRNAs$category = "Exclusive_lncRNA"
candidate_lncRNAs_common$category = "Common_lncRNA"

biologically_relevant_lncRNAs <- rbind(exclusive_biologically_relevant_lncRNAs, candidate_lncRNAs_common,exclusive_lncRNAs_CRISPRi)
head(biologically_relevant_lncRNAs)
```







